File: 01_GOF23/DP_examples/CSharp/02_Structural/Facade/Facade.cs
namespace Structural.Facade;

public enum ShapeType
{
    Circle,
    Rectangle,
    Square
}

public class ShapeMaker
{
    public static readonly ShapeMaker Instance = new ShapeMaker();
    public IShape BuildShape(ShapeType shapeType, params int[] args)
    {
        switch (shapeType)
        {
            case ShapeType.Circle:
                return new Circle(args[0]);
            case ShapeType.Rectangle:
                return new Rectangle(args[0], args[1]);
            case ShapeType.Square:
                return new Square(args[0]);
            default:
                return null;
        }
    }
}

File: 01_GOF23/DP_examples/Lua/02_Structural/Facade/Facade.lua
-- 加载 Subsystem 模块
require("Subsystem")

-- 形状类型枚举
---@enum ShapeType
ShapeType = {
    CIRCLE = 0,
    RECTANGLE = 1,
    SQUARE = 2
}

-- 形状创建者（外观类）
---@param shapeType ShapeType
---@param ... number[]
---@return IShape
function BuildShape(shapeType, ...)
    local args = {...}
    if shapeType == ShapeType.CIRCLE then
        if #args >= 1 then
            return Circle:new(args[1])
        end
    elseif shapeType == ShapeType.RECTANGLE then
        if #args >= 2 then
            return Rectangle:new(args[1], args[2])
        end
    elseif shapeType == ShapeType.SQUARE then
        if #args >= 1 then
            return Square:new(args[1])
        end
    end
    return nil
end
